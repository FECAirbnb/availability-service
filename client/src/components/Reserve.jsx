/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/self-closing-comp */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable react/prop-types */
/* eslint-disable no-unused-vars */
/* eslint-disable import/extensions */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/button-has-type */
/* eslint-disable react/jsx-one-expression-per-line */
import React from 'react';
import Calendar from './Calendar.jsx';
// import AddGuests from './AddGuests.jsx';

class Reserve extends React.Component {
  constructor(props) {
    super(props);
    this.container = React.createRef();
    this.state = {
      open: false,
      clickGuest: false,
      checkIn: 'Check in',
      checkOut: 'Check out',
      guestCount: 1
    };

    this.handleDropdownClick = this.handleDropdownClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.handleDateChange = this.handleDateChange.bind(this);
    this.handleStatechange = this.handleStatechange.bind(this);
    this.handleGuestCount = this.handleGuestCount.bind(this);
    this.handleGuestDropdown = this.handleGuestDropdown.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOutsideClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOutsideClick);
  }

  handleDropdownClick() {
    this.setState(state => {
      return {
        open: true
      };
    });
  }

  handleGuestDropdown() {
    if (!this.state.clickGuest) {
      this.setState(state => {
        return {
          clickGuest: true
        };
      });
    } else {
      this.setState(state => {
        return {
          clickGuest: false
        };
      });
    }
  }

  handleOutsideClick(e) {
    if (this.container.current && !this.container.current.contains(e.target)) {
      this.setState({
        open: false,
        clickGuest: false
      });
    }
  }

  handleDateChange(date) {
    if (this.state.checkIn === 'Check in') {
      this.setState({
        checkIn: date
      });
    } else {
      this.setState({
        checkOut: date
      });
    }
  }

  handleStatechange() {
    this.setState({
      open: false,
      checkIn: 'Check in',
      checkOut: 'Check out'
    });
    this.props.onClick();
  }

  handleGuestCount() {
    let guestStr = '';
    if (this.state.guestCount === 1) {
      guestStr = 'guest';
    } else {
      guestStr = 'guests';
    }
    return guestStr;
  }

  handleArrowChange() {
    let arrow = '';
    if (this.state.clickGuest === false) {
      arrow = (
        <img
          src="https://image.flaticon.com/icons/png/128/566/566015.png"
          data-src="https://image.flaticon.com/icons/png/128/566/566015.png"
          srcSet="https://image.flaticon.com/icons/svg/566/566015.svg 4x"
          alt="Down chevron free icon"
          title="Down chevron free icon"
          width="64"
          height="64"
          className="arrow"
        ></img>
      );
    } else {
      arrow = (
        <img
          src="https://image.flaticon.com/icons/png/128/566/566014.png"
          data-src="https://image.flaticon.com/icons/png/128/566/566014.png"
          srcSet="https://image.flaticon.com/icons/svg/566/566014.svg 4x"
          alt="Up chevron free icon"
          title="Up chevron free icon"
          width="64"
          height="64"
          className="arrow"
        ></img>
      );
    }
    return arrow;
  }

  render() {
    return (
      <div className="reserve" ref={this.container}>
        <div>
          <span className="price">${this.props.state.data[0].price}</span>
          <span className="per-night"> per night</span>
        </div>
        <div>
          <span role="img" aria-label="Rating 1 out of 5" className="_m5wjo84">
            4.99
          </span>
          <span className="per-night">(6 reviews)</span>
        </div>
        <div className="date-picker">
          <div id="check-in" onClick={this.handleDropdownClick}>
            {this.state.checkIn}
          </div>

          <div id="check-out" onClick={this.handleDropdownClick}>
            {this.state.checkOut}
          </div>
        </div>
        <div>
          {this.state.open && (
            <div className="dropdown">
              <Calendar state={this.props.state} handleDateChange={this.handleDateChange} />
            </div>
          )}
        </div>
        <div className="add-guest">
          <div id="guest-count" onClick={this.handleGuestDropdown}>
            {this.state.guestCount} {this.handleGuestCount()}
          </div>

          <div id="arrow" onClick={this.handleGuestDropdown}>
            {this.handleArrowChange()}
          </div>
        </div>
        <div>
          {this.state.clickGuest && (
            <div className="dropdown">
              <div>
                <div>hello</div>
              </div>
            </div>
          )}
        </div>
        <div>
          <button id="reserve-btn" type="submit" onClick={this.handleStatechange}>
            RESERVE
          </button>
        </div>
      </div>
    );
  }
}

export default Reserve;
