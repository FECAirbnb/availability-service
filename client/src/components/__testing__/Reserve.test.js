/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import Enzyme, { shallow, render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import Reserve from '../Reserve';

Enzyme.configure({ adapter: new Adapter() });

const sampleLoc = {
  state: {
    data: [
      {
        id: 1,
        name: 'Bora bora',
        price: 1450,
        cleaning_fee: 100,
        service_fee: 100,
        date: '2020-03-01T07:00:00.000Z',
        Location_id: 1,
        Dates_id: 1
      }
    ]
  }
};
describe('Reserve component', () => {
  it('renders correctly enzyme', () => {
    const wrapper = shallow(<Reserve state={sampleLoc.state} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a location number as a state', () => {
    const wrapper = shallow(<Reserve state={sampleLoc.state} />);
    expect(wrapper.state('open')).toEqual(false);
  });

  it('should add to the guest count state when clicked', () => {
    const wrapper = shallow(<Reserve state={sampleLoc.state} />);
    setTimeout(() => {
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(2);
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(3);
      done();
    }, 1000);
  });

  it('should subtract from the guest count state when clicked', () => {
    const wrapper = shallow(<Reserve state={sampleLoc.state} />);
    setTimeout(() => {
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(0);
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(1);
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(2);
      wrapper.find('button').simulate('click');
      expect(wrapper.state('guestCount')).toEqual(1);
      done();
    }, 1000);
  });
});
