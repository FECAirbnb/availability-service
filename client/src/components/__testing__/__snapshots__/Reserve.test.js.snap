// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reserve component renders correctly enzyme 1`] = `
<div
  className="reserve"
>
  <div>
    <span
      className="price"
    >
      $
      1450
    </span>
    <span
      className="per-night"
    >
       per night
    </span>
  </div>
  <div>
    <img
      alt="Star icon"
      className="star"
      src="https://img.icons8.com/material-sharp/2x/star.png"
    />
    <span
      className="rating"
    >
      4.99
    </span>
    <span
      className="per-night"
    >
      (6 reviews)
    </span>
  </div>
  <div
    className="date-picker"
  >
    <div
      id="check-in"
      onClick={[Function]}
    >
      Check in
    </div>
    <div
      id="check-out"
      onClick={[Function]}
    >
      Check out
    </div>
  </div>
  <div />
  <div
    className="add-guest"
  >
    <div
      id="guest-count"
      onClick={[Function]}
    >
      1
       
      guest
    </div>
    <div
      id="arrow"
      onClick={[Function]}
    >
      <img
        alt="Down chevron free icon"
        className="arrow"
        data-src="https://image.flaticon.com/icons/png/128/566/566015.png"
        height="64"
        src="https://image.flaticon.com/icons/png/128/566/566015.png"
        srcSet="https://image.flaticon.com/icons/svg/566/566015.svg 4x"
        title="Down chevron free icon"
        width="64"
      />
    </div>
  </div>
  <div />
  <div />
  <div>
    <button
      id="reserve-btn"
      onClick={[Function]}
      type="submit"
    >
      RESERVE
    </button>
  </div>
</div>
`;
